source/opt/passes.h

```diff
@@ -74,6 +79,132 @@ class EliminateDeadConstantPass : public Pass {
   bool Process(ir::Module*) override;
 };
 
+// The optimization pass for folding SpecConstantOp and SpecConstantComposite
+// instructions. After freezing, spec constants defined with Spec ID are
+// specialized into front-end constants. This pass folds the SpecConstantOp
+// instructions and SpecConstantComposite instructions that depend on those
+// frozen spec constants.
+class FoldSpecConstantOpAndCompositePass : public Pass {
+ public:
+   // Data structure to record the in-operand of front-end constants.
+   // ScalarValue stores the value of of scalar constants. CompositeValue
+   // stores a vector of its component ids. Value is a abstract class of
+   // ScalarValue and CompositeValue.
+  struct ScalarValue;
+  struct CompositeValue;
+  struct Value {
+    virtual ScalarValue* AsScalar() { return nullptr; }
+    virtual CompositeValue* AsComposite() { return nullptr; }
+  };
+  struct ScalarValue : public Value {
+    ScalarValue(uint32_t w) : word(w) {}
+    ScalarValue(const ScalarValue* other) : word(other->word) {}
```

Comment:
The convention is to use `const <type>&` instead of `const <type>*` for copy constructors.